res/controllers/Numark-Scratch-scripts.js

```diff
@@ -0,0 +1,423 @@
+            this.group = "[Library]";
+            this.inKey = "MoveUp";
+        }
+    });
+
+    this.loopButton = new components.Button({
+        midi: [0x93 + channel, 0x3F],
+        [encoderInvert("unshift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "beatloop_activate";
+        },
+        [encoderInvert("shift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "LoadSelectedTrack";
+        }
+    });
+
+    this.padSection = new NumarkScratch.PadSection(number);
+
+    this.shiftButton = new components.Button({
+        midi: [0x9F, 0x32],
+        input: function(channel, control, value) {
+            if (this.isPress(channel, control, value)) {
+                NumarkScratch.shift();
+                this.send(this.on);
+            } else {
+                NumarkScratch.unshift();
+                this.send(this.off);
+            }
+        }
+    });
+
+    this.reconnectComponents(function(component) {
+        if (component.group === undefined) {
+            component.group = this.currentDeck;
+        }
+    });
+};
+
+NumarkScratch.Deck.prototype = new components.Deck();
+
+// Pad modes control codes
+NumarkScratch.PadModeControls = {
+    HOTCUE: 0x00,
+    SAMPLER: 0x07,
+    ROLL: 0x0B,
+};
+
+NumarkScratch.PadSection = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.modes = {};
+    this.modes[NumarkScratch.PadModeControls.HOTCUE] = new NumarkScratch.ModeHotcue(deckNumber);
+    this.modes[NumarkScratch.PadModeControls.SAMPLER] = new NumarkScratch.ModeSampler(deckNumber);
+    this.modes[NumarkScratch.PadModeControls.ROLL] = new NumarkScratch.ModeRoll(deckNumber);
+
+    this.modeButtonPress = function(channel, control, _value) {
+        this.setMode(channel, control);
+    };
+
+    this.padPress = function(channel, control, value, status, group) {
+        const i = (control - 0x14) % 8;
+        this.currentMode.pads[i].input(channel, control, value, status, group);
+    };
+
+    this.setMode = function(_channel, control) {
+        const newMode = this.modes[control];
+        this.currentMode.forEachComponent(function(component) {
+            component.disconnect();
+        });
+
+        newMode.forEachComponent(function(component) {
+            component.connect();
+            component.trigger();
+        });
+
+        this.currentMode = newMode;
+    };
+
+    this.currentMode = this.modes[NumarkScratch.PadModeControls.HOTCUE];
+};
+NumarkScratch.PadSection.prototype = Object.create(components.ComponentContainer.prototype);
+
+NumarkScratch.ModeHotcue = function(deckNumber) {
+    components.ComponentContainer.call(this);
+
+    this.control = NumarkScratch.PadModeControls.HOTCUE;
+
+    this.pads = new components.ComponentContainer();
+    for (let i = 0; i < 4; i++) {
+        this.pads[i] = new components.HotcueButton({
+            group: "[Channel" + deckNumber + "]",
```

Comment:
these could be template-literals everywhere.
```suggestion
            group: `[Channel${deckNumber}]`,
```

Reply:
done